src/transaction/buffer.rs

```diff
@@ -163,10 +191,14 @@ impl Buffer {
 
     /// Converts the buffered mutations to the proto buffer version
     pub async fn to_proto_mutations(&self) -> Vec<kvrpcpb::Mutation> {
-        self.mutations
-            .lock()
-            .await
+        let mutations = self.mutations.lock().await;
+        let (primary, other) = mutations
+            .key_mutation_map
             .iter()
+            .partition::<Vec<_>, _>(|(key, _)| *key == mutations.primary_key.as_ref().unwrap());
+        primary
+            .into_iter()
+            .chain(other.into_iter())
             .filter_map(|(key, mutation)| mutation.to_proto_with_key(key))
             .collect()
```

Comment:
This is really ugly, do we really need these mutations sorted? Or we can replace the `BTreeMap` with a naive `Vec`.

Reply:
I think we need to keep the BTreeMap because we look up writes in our cache before every read, which could be expensive if the data is not sorted.

I think we should either not promise that the primary key is the first element here, or we could have a more complex data structure, like using a Vec of mutations and a BTree index of the keys.